src/platform/backends/hyperv_api/hcn/hyperv_hcn_api_wrapper.cpp

```diff
@@ -0,0 +1,353 @@
+    // function.
+    UniqueCotaskmemString result_msgbuf{};
+
+    // Perform the operation. The last argument of the all HCN operations (except
+    // HcnClose*) is ErrorRecord, which is a JSON-formatted document emitted by
+    // the API describing the error happened. Therefore, we can streamline all API
+    // calls through perform_hcn_operation.
+    const auto result =
+        ResultCode{fn(std::forward<Args>(args)..., out_ptr(result_msgbuf, api.CoTaskMemFree))};
+
+    mpl::debug(kLogCategory, "perform_hcn_operation(...) > result: {}", static_cast<bool>(result));
+
+    // Error message is only valid when the operation resulted in an error.
+    // Passing a nullptr is well-defined in "< C++23", but it's going to be
+    // forbidden afterwards. Going an extra mile just to be future-proof.
+    return {result, {result_msgbuf ? result_msgbuf.get() : L""}};
+}
+
+// ---------------------------------------------------------
+
+/**
+ * Open an existing Host Compute Network and return a handle to it.
+ *
+ * This function is used for altering network resources, e.g. adding a new
+ * endpoint.
+ *
+ * @param api The HCN API table
+ * @param network_guid GUID of the network to open
+ *
+ * @return UniqueHcnNetwork Unique handle to the network. Non-nullptr when successful.
+ */
+UniqueHcnNetwork open_network(const HCNAPITable& api, const std::string& network_guid)
+{
+    mpl::debug(kLogCategory, "open_network(...) > network_guid: {} ", network_guid);
+
+    UniqueHcnNetwork network{};
+    const auto result = perform_hcn_operation(api,
+                                              api.OpenNetwork,
+                                              guid_from_string(network_guid),
+                                              out_ptr(network, api.CloseNetwork));
+    if (!result)
+    {
+        mpl::error(kLogCategory, "open_network() > HcnOpenNetwork failed with {}!", result.code);
+    }
+    return network;
+}
+
+/**
+ * Determine the log severity level for a HCN error.
+ *
+ * @param [in] result Operation result
+ * @return mpl::Level The determined severity level
+ */
+auto hcn_errc_to_log_level(const OperationResult& result)
+{
+    /**
+     * Some of the errors are "expected", e.g. a network may be already
+     * exist and that's not necessarily an error.
+     */
+    switch (static_cast<HRESULT>(result.code))
+    {
+    case HCN_E_NETWORK_ALREADY_EXISTS:
+        return mpl::Level::debug;
+    }
+
+    return mpl::Level::error;
+}
+
+} // namespace
+
+// ---------------------------------------------------------
+
+HCNWrapper::HCNWrapper(const HCNAPITable& api_table) : api{api_table}
+{
+    mpl::debug(kLogCategory, "HCNWrapper::HCNWrapper(...): api_table: {}", api);
+}
+
+// ---------------------------------------------------------
+
+OperationResult HCNWrapper::create_network(const CreateNetworkParameters& params) const
+{
+    mpl::debug(kLogCategory, "HCNWrapper::create_network(...) > params: {} ", params);
```

Comment:
idem: this feels redundant with respect to the log in `perform_hcn_operation`. I realize the args aren't logged in there, but they could if that's the goal, right?